Assets/HoloToolkit/Input/Scripts/Gaze/GazeManager.cs

```diff
@@ -294,25 +295,37 @@ private void RaycastUnityUI()
         private RaycastResult FindClosestRaycastHitInLayerMasks(List<RaycastResult> candidates, LayerMask[] layerMaskList)
         {
             int combinedLayerMask = 0;
-            for (int i = 0; i < layerMaskList.Length; i++)
+            for (var i = 0; i < layerMaskList.Length; i++)
             {
                 combinedLayerMask = combinedLayerMask | layerMaskList[i].value;
             }
 
-            RaycastResult? minHit = null;
-            for (var i = 0; i < candidates.Count; ++i)
+            candidates = candidates.Where(c => c.gameObject != null).Where(c => IsLayerInLayerMask(c.gameObject.layer, combinedLayerMask)).ToList();
+            candidates.Sort(CompareRaycasts);
+            return candidates.DefaultIfEmpty(new RaycastResult()).FirstOrDefault();
+        }
+
+        private static int CompareRaycasts(RaycastResult left, RaycastResult right)
+        {
+            var result = CompareRaycastsByCanvasDepth(left, right);
+            if (result != 0) { return result; }
+            return CompareRaycastsByDistance(left, right);
+        }
+
+        private static int CompareRaycastsByCanvasDepth(RaycastResult left, RaycastResult right)
+        {
+            var canvas1 = left.gameObject.GetComponentInParent<Canvas>();
+            var canvas2 = right.gameObject.GetComponentInParent<Canvas>();
```

Comment:
Is there a way to do this without the need to get the canvas component?

Reply:
I need the canvas. I guess I could make a container to hold both to cache it.

Reply:
We just need to know if both raycast results are from the same canvas?

Reply:
Exactly